cms/src/builder/kari.rs

```diff
@@ -0,0 +1,446 @@
+        OriginatorIdentifierOrKey, OriginatorPublicKey, RecipientEncryptedKey, RecipientInfo,
+    },
+};
+
+// Internal imports
+use const_oid::{AssociatedOid, ObjectIdentifier};
+use der::{
+    asn1::{BitString, OctetString},
+    Any, Decode, Encode, Sequence,
+};
+
+// Alloc imports
+use alloc::{vec, vec::Vec};
+
+// RustCrypto imports
+use elliptic_curve::{
+    ecdh::EphemeralSecret,
+    point::PointCompression,
+    sec1::{FromEncodedPoint, ModulusSize, ToEncodedPoint},
+    AffinePoint, CurveArithmetic, FieldBytesSize, PublicKey,
+};
+
+/// The `EccCmsSharedInfo` type is defined in [RFC 5753 Section 7.2].
+///
+/// ```text
+///   EccCmsSharedInfo ::= SEQUENCE {
+///       keyInfo         AlgorithmIdentifier,
+///       entityUInfo [0] EXPLICIT OCTET STRING OPTIONAL,
+///       suppPubInfo [2] EXPLICIT OCTET STRING  }
+/// ```
+///
+/// [RFC 5753 Section 7.2]: https://www.rfc-editor.org/rfc/rfc5753#section-7.2
+#[derive(Clone, Debug, Eq, PartialEq, Sequence)]
+pub struct EccCmsSharedInfo {
+    /// Object identifier of the key-encryption algorithm
+    pub key_info: AlgorithmIdentifierOwned,
+    /// Additional keying material - optional
+    #[asn1(
+        context_specific = "0",
+        tag_mode = "EXPLICIT",
+        constructed = "true",
+        optional = "true"
+    )]
+    pub entity_u_info: Option<UserKeyingMaterial>,
+    /// Length of the generated KEK, in bits, represented as a 32-bit number
+    #[asn1(context_specific = "2", tag_mode = "EXPLICIT", constructed = "true")]
+    pub supp_pub_info: OctetString,
+}
+
+/// Represents supported key agreement algorithm for ECC - as defined in [RFC 5753 Section 7.1.4].
+///
+/// As per [RFC 5753 Section 8]:
+/// ```text
+/// Implementations that support EnvelopedData with the ephemeral-static
+/// ECDH standard primitive:
+///
+/// - MUST support the dhSinglePass-stdDH-sha256kdf-scheme key
+///    agreement algorithm, the id-aes128-wrap key wrap algorithm, and
+///    the id-aes128-cbc content encryption algorithm; and
+/// - MAY support the dhSinglePass-stdDH-sha1kdf-scheme, dhSinglePass-
+///    stdDH-sha224kdf-scheme, dhSinglePass-stdDH-sha384kdf-scheme, and
+///    dhSinglePass-stdDH-sha512kdf-scheme key agreement algorithms;
+///    the id-alg-CMS3DESwrap, id-aes192-wrap, and id-aes256-wrap key
+///    wrap algorithms; and the des-ede3-cbc, id-aes192-cbc, and id-
+///    aes256-cbc content encryption algorithms; other algorithms MAY
+///    also be supported.
+/// ```
+///
+/// As such the following are currently supported:
+/// - dhSinglePass-stdDH-sha224kdf-scheme
+/// - dhSinglePass-stdDH-sha256kdf-scheme
+/// - dhSinglePass-stdDH-sha384kdf-scheme
+/// - dhSinglePass-stdDH-sha512kdf-scheme
+///
+/// [RFC 5753 Section 7.1.4]: https://datatracker.ietf.org/doc/html/rfc5753#section-7.1.4
+/// [RFC 5753 Section 8]: https://datatracker.ietf.org/doc/html/rfc5753#section-8
+#[allow(clippy::enum_variant_names)]
+#[derive(Clone, Copy)]
+pub enum KeyAgreementAlgorithm {
```

Comment:
Could we convert that to a trait instead:
```
pub trait KeyAgreementAlgorithm: AssociatedOid {
    fn kdf(secret: &[u8]);
}
```

And then make a struct like:
```
struct DhSinglePassStdDhKdf<D> where D:Digest + {
    digest: PhantomData<D>,
}

impl<D> KeyAgreementAlgorithm for DhSinglePassStdDhKdf<D> where D:Digest {
    fn kdf(secret: &[u8]) {
        ansi_x963_kdf::derive_key_into::<D>(secret)
    }
}

impl AssociatedOid for DhSinglePassStdDhKdf<sha2::Sha224> {
   const OID: ObjectIdentifier = rfc5753::DH_SINGLE_PASS_STD_DH_SHA_224_KDF_SCHEME;
}
```

Reply:
I had something like that in mind:
https://github.com/nemynm/formats/pull/2

Reply:
Thanks!
Indeed I was not particularly convinced by the current KDF handling, but I thought it was a fair solution that was easily extensible for other schemes/algo (maybe with the help of a quick macro to derive `try_ansi_x963_kdf`).

At first I actually went the trait route in my code (close-ish to what you proposed), but then settled on the enum and dispatch function. Rationales were:
1. This snippet in `EnvelopedDataBuilder` at [builder.rs#L890](https://github.com/RustCrypto/formats/blob/master/cms/src/builder.rs#L890) which caught my eye:
```
    // TODO bk Not good to offer both, `content_encryptor` and `content_encryption_algorithm`.
    // We should
    // (1) either derive `content_encryption_algorithm` from `content_encryptor` (but this is not
    //            yet supported by RustCrypto),
    // (2) or     pass `content_encryption_algorithm` and create an encryptor for it.
    // In the first case, we might need a new trait here, e.g. `DynEncryptionAlgorithmIdentifier` in
    // analogy to `DynSignatureAlgorithmIdentifier`.
    // Going for (2)
    //  content_encryptor: E,
    content_encryption_algorithm: ContentEncryptionAlgorithm,
```

Even if its not the same situation, I interprated it as "work with enum rather than trait", at least until some is finalized. That was somewhat re-inforced by the current [get_hasher()](https://github.com/RustCrypto/formats/blob/master/cms/src/builder.rs#L1061) function.

2. Consequently, just passing an enum looked closer to the current API for other builders (having to specify the KA generic in `KeyAgreeRecipientInfoBuilder` seemed a tad less user-friendly).

3. I thought that if we ended up supporting the other key agreement algorithm (e.g. cofactor ECDH and 1-Pass ECMQV ), we may want to have a more complete trait like:
```
pub trait KeyAgreementAlgorithm: AssociatedOid {
    fn shared_secret()
    fn try_kdf(secret: &[u8], other_info: &[u8], key: &mut impl AsMut<[u8]>) -> Result<()>;
}
```
-> So essentially folding most of the `build()` logic behind a trait - so to speak.

However not exactly knowing yet:
- what `shared_secret()` signature should look like for all.
- the input to `try_kdf()` would be with other schemes (would that always be a `SharedSecret<C>` or something more `generic` like `&[u8]`).
- if we should have three structs (`DhSinglePassStdDhKdf`, `DhSinglePassCoFactorDhKdf`, `MqvSinglePassKdf`) or maybe just one generic for the three schemes.
 
I figured that a trait was maybe not the best solution yet.

Anyway as you can see, these are more generic thoughts than strong reasons not to, so I am also ok to go the trait route now too.